availability-store/src/lib.rs

```diff
@@ -196,76 +249,334 @@ impl Store {
 			}
 		}
 	}
+
+
+	pub fn erasure_chunk(&self, relay_parent: Hash, candidate_hash: Hash, chunk_id: usize) -> Option<ErasureChunk> {
+		let encoded_key = erasure_chunks_key(&relay_parent, &candidate_hash);
+
+		let chunks = match self.inner.get(columns::DATA, &encoded_key[..]) {
+			Ok(Some(raw)) => ErasureChunks::decode(&mut &raw[..]).expect("all stored data serialized correctly; qed"),
+			Ok(None) => return None,
+			Err(e) => {
+				warn!(target: "availability", "Error reading from availability store: {:?}", e);
+				return None;
+			}
+		};
+
+		chunks.chunks.iter().find_map(|c| {
+			match c.index as usize == chunk_id {
+				true => Some(c.clone()),
+				false => None,
+			}
+		})
+	}
+
+	fn try_recover_block_data_and_messages(
+		&self,
+		relay_parent: &Hash,
+		candidate_hash: &Hash
+	) -> Option<(BlockData, Option<AvailableMessages>)> {
+		let encoded_key = erasure_chunks_key(&relay_parent, &candidate_hash);
+
+		let chunks = match self.inner.get(columns::DATA, &encoded_key[..]) {
+			Ok(Some(raw)) => ErasureChunks::decode(&mut &raw[..]).expect("all stored data serialized correctly; qed"),
+			Ok(None) => return None,
+			Err(e) => {
+				warn!(target: "availability", "Error reading from availability store: {:?}", e);
+				return None;
+			}
+		};
+
+		let chunks_ref: Vec<_> = chunks.chunks.iter()
+			.map(|ErasureChunk { chunk, index, .. }| (&chunk[..], (*index as usize)))
+			.collect();
```

Comment:
the collect here is extraneous, as far as I can see, since `erasure::reconstruct` accepts an `IntoIterator`.

you might have to create another internal scope to avoid borrow checker issues.
```rust
let res = {
   let chunks_iter = ...; // no collect
   erasure::reconstruct(chunks.n_validators, chunks_ref)
}
```